cpp/ql/src/Security/CWE/CWE-611/XXE.ql

```diff
@@ -0,0 +1,203 @@
+/**
+ * @name External Entity Expansion
+ * @description Parsing user-controlled XML documents and allowing expansion of
+ *              external entity references may lead to disclosure of
+ *              confidential data or denial of service.
+ * @kind path-problem
+ * @id cpp/external-entity-expansion
+ * @problem.severity warning
+ * @security-severity 9.1
+ * @precision medium
+ * @tags security
+ *       external/cwe/cwe-611
+ */
+
+import cpp
+import semmle.code.cpp.ir.dataflow.DataFlow
+import DataFlow::PathGraph
+import semmle.code.cpp.ir.IR
+
+/**
+ * A flow state representing a possible configuration of an XML object.
+ */
+abstract class XXEFlowState extends string {
```

Comment:
Is there a particular reason you use `extends string` here and not `extends DataFlow::FlowState`?

Reply:
You're right, `extends DataFlow::FlowState` is better (updated).